kyo-direct/shared/src/test/scala/kyo/ShiftTest.scala

```diff
@@ -57,244 +61,404 @@ class ShiftTest extends AnyFreeSpec with Assertions:
+                            Choice.eval(List(true, false)).map(b =>
+                                if b then str.toUpperCase else str
+                            )
+                        }
+                    val result = Choice.run(defer(effects.map(_.now))).eval
+
+                    assert(result.contains(Chunk("X", "Y")))
+                    assert(result.contains(Chunk("X", "y")))
+                    assert(result.contains(Chunk("x", "Y")))
+                    assert(result.contains(Chunk("x", "y")))
+                    assert(result.size == 4)
+                }
+
+                "flatMap" in {
+                    val effects = defer:
+                        val tf = Choice.eval(List(true, false)).later
+                        Coll("x", "y").flatMap(str1 =>
+                            val pred1 = tf.now
+                            Coll("z").map(str2 =>
+                                val pred2 = tf.now
+                                if pred1 & pred2 then
+                                    (str1 + str2).toUpperCase
+                                else if pred1 != pred2 then
+                                    Choice.drop.now
+                                else (str1 + str2)
+                                end if
+                            )
+                        )
+
+                    val result = Choice.run(effects).eval
+                    assert(result.contains(Chunk("XZ", "YZ")))
+                    assert(result.contains(Chunk("XZ", "yz")))
+                    assert(result.contains(Chunk("xz", "YZ")))
+                    assert(result.contains(Chunk("xz", "yz")))
+                    assert(result.size == 4)
+                }
+
+                "foldLeft" in {
+                    val result = Choice.run(
+                        defer:
+                            Coll(1, 1).foldLeft(0)((acc, _) => Choice.eval(List(0, 1)).map(n => acc + n).now)
+                    ).eval
+
+                    assert(result.contains(0))
+                    assert(result.contains(1))
+                    assert(result.contains(2))
+                    assert(result.size == 4)
+                }
+            }
         }
-        /*"reduceOption" in {
-            val d: Option[Int] < Any = defer:
-                xs.reduceOption(_.now + _.now).map(_.now)
-
-            assert(d.eval == Option(10))
-        }*/
-        "map" in {
-            val d: Seq[Int] < Any = defer:
-                xs.map(_.now).toSeq
+    end iterableTests
 
-            assert(d.eval == Seq(1, 2, 3, 4))
-        }
-        "flatMap" in {
-            val d: Seq[Int] < Any = defer:
-                def f(i: Int) = Seq(1)
-                xs.flatMap(i => f(i.now)).toSeq.distinct
+    iterableTests[Iterable]("Iterable", [X] => seq => seq)
+    iterableTests[Vector]("Vector", [X] => seq => Vector.from(seq))
+    iterableTests[Chunk]("Chunk", [X] => seq => Chunk.from(seq))
+    iterableTests[List]("List", [X] => seq => List.from(seq))
```

Comment:
@hearnadam it is generalized on the collection structures, but not on the implicit selection from `dotty-cps-async`.
So far inlines are not solving it.